### ISSUE TITLE
Can we improve the output of npmgraph?

### ISSUE THREAD

**Issue by fregante** [npmgraph/npmgraph#93]
- https://github.com/pahen/madge/issues/289

This impacts npmgraph, but to a lesser extent because the graph you chose is slightly better in my experience.

**Comment by broofa** [comment]
NPMGraph has a couple of tools that help with this.  Namely the highlighting to direct dependencies upon selection, and the ability to collapse nodes gives users a bit more control over graph complexity.  Granted, this isn't necessarily the ideal solution but I don't think a tree-view UI (ala the MacOS file "Finder" interface) is a good solution, either.  Those sorts of UIs aren't really designed for dealing with cyclic graphs.  They don't make it easy to see shared or cyclic dependencies.  Nor do they give a good birds-eye view of the entire graph in the way the current solution does.

And if you want a tree view of dependencies, that's basically what `npm ls` provides.

![image](https://user-images.githubusercontent.com/164050/126051395-c4ee0c94-ea31-4fda-aa59-76c3c999a18f.png)

**Comment by fregante** [comment]
> the highlighting to direct dependencies upon selection

Yes that helps with medium-sized graphs

> Those sorts of UIs aren't really designed for dealing with cyclic graphs

Yes they are, this is a cyclic dependency:

```
mkdir -p a/b
cd a/b
ln -s .. a
```

<img width="1716" alt="Screen Shot" src="https://user-images.githubusercontent.com/1402241/126051721-da8c2319-12ea-4d23-ae59-1ce7b5fc001e.png">

It works just fine, it just depends on what the intent is. That brings me to 👇 

> Nor do they give a good birds-eye view of the entire graph in the way the current solution does.

My intent is to explore dependencies level by level and that type of graph is just hard to follow when there are many dependencies. Example: https://npmgraph.js.org/?q=jest

> And if you want a tree view of dependencies, that's basically what `npm ls` provides.

That's a good point! Indeed for npmgraph that output isn't as helpful since `npm ls` already does it, like: `npm ls --level=0`

**Comment by broofa** [comment]
> > Those sorts of UIs aren't really designed for dealing with cyclic graphs

> Yes they are, this is a cyclic dependency...

Tree views allow you to navigate such dependencies, as you've shown, _but only by pretending they're not actually cyclic_.  The finder image you provide gives no indication that what you're looking at is a cyclic path.  It could just as easily be a 6 or 7 level deep directory structure.  Hence why I say they're not _designed_ for cyclic dependencies.

This is because links are relatively rare in filesystems, and most users would find any UI that reflects their true nature all the more confusing.  This is distinctly not the case for module dependency graphs.  The nature and directionality of cyclic dependencies in NPM is primary concern of our users.

Note, too, that this whitewashing of cyclic links is particularly problematic if you're interested in shared ancestry.  E.g.:

```
mkdir -p a a/b
cd a
ln -s b c
ln -s b d
```
![image](https://user-images.githubusercontent.com/164050/126052569-847d8180-d810-4ead-8271-6bf72bb417b0.png)

There's simply no way of knowing that c & d have b as a common ancestor.

> My intent is to explore dependencies level by level and that type of graph is just hard to follow when there are many dependencies

I agree that navigation is challenging with large graphs.  `jest` is a great example.  Any sort of "birds-eye" view suffers from information overload in these cases.  (E.g. https://npm.anvaka.com/#/view/2d/jest is... well... no comment needed.)

----

Thinking outside the box for a moment, what about a UI that looked something like this:
![image](https://user-images.githubusercontent.com/164050/126053072-a3e40d03-8c98-44dc-88ec-966e9d7ff441.png)

I.e. A breadcrumb path at the top that shows you the dependency chain you've clicked through, and a closeup dependency graph of whatever module you're currently focused on below that, with immediate parents to the left of the focused module and immediate children to the right.

Clicking on a module in the graph navigate to that module.  _However_, if you click on a module that's already in your breadcrump path, it leaves the path unaltered, but focuses on the appropriate module in the path.  E.g. if you were to click on `cliui`, above, you'd see the appropiate closeup diagram, with this breadcrumb path:
![image](https://user-images.githubusercontent.com/164050/126053191-87ddc338-41ea-409d-9bf7-804d289a3ee2.png)

I.e. you'd still be able to click on `wrap-ansi` or `string-width` to return to points deeper in the path you've been exploring. 

This would avoid the infinite-recursion problem that cyclic dependencies present in treeviews, while also allowing users to easily navigate back and forth among modules in a path without losing their "place".

**Comment by fregante** [comment]
> There's simply no way of knowing that c & d have b as a common ancestor.

Simply put: I don't care. To reiterate: It just depends on what one needs to do. If I were trying to navigate Jest’s dependencies, cyclic dependencies just make the tool completely useless as you've seen. I'd rather _not_ know about common ancestors.

This is not to say that npmgraph is broken, I'm just saying that it's not one-size-fits-all(-use-cases).

> E.g. [npm.anvaka.com/#/view/2d/jest](https://npm.anvaka.com/#/view/2d/jest) is... well... no comment needed.)

heh yeah that tool is only useful to demonstrate the mess of npm dependencies, which is how I found npmgraph: https://github.com/npmhub/npmhub/pull/87

> A breadcrumb path at the top

That sounds great, but I suppose you have to click each step along the way or else if you just click on `E`:

```
A- - - - - - E
  \         /
   B - C - D
```

You'll end up with 2 paths: (or more in other packages)

```
A > E
A > B > C > D > E
```

**Comment by broofa** [comment]
> That sounds great, but I suppose you have to click each step along the way...

Don't think of the breadcrumb path as the path back to the root node in the larger graph.  That's problematic for all sorts of reasons.  The breadcrumb path is just a record of which modules you've clicked once you arrive at this view (let's call it the "Detail Graph").

How you arrive at this view, is unimportant.  The current UI has lots of ways in which to make that happen.  E.g. double-click on a module in the graph, or anywhere in the inspector?  Whatever.  The point is, when you open the Detail Graph up, the breadcrumb path is just the name of the current module.

The path only changes as you navigate around within the Detail Graph.

This is important because the way in which you navigate may result in a path that's rather unintuitive if you assume it should reflect the path to the root node.  E.g. If you start at "E" and click up the ancestor chain, you get a path of "E > D > C > B > A".

Aside: I probably should have labeled my mockup, above, with "Upstream" and "Downstream" rather than "Parents" and "Children".  In cyclic graphs, the parent-child relationship can be a bit nonsensical.  E.g. In a graph of "[A -> B -> C -> A](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%20%20rankdir%3DLR%0A%20%20B%20-%3E%20C%20-%3E%20A%20-%3E%20B%0A%7D)", there's no obvious "parent" for the graph.

**Comment by broofa** [comment]
[Riffing on the DOT notation for the detail graph](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%20%20rankdir%3DLR%0A%20%20node%20%5Bshape%3Dbox%20style%3Drounded%20fontname%3D%22sans-serif%22%5D%3B%0A%0A%20%20subgraph%20cluster_0%20%7B%0A%20%20%20%20label%20%3D%20%22Upstream%22%3B%0A%20%20%20%20fontname%3D%22sans-serif%22%3B%0A%20%20%20%20fontcolor%3Dgrey%3B%0A%20%20%20%20shape%3Dbox%3B%0A%20%20%20%20style%3Drounded%3B%0A%0A%20%20%20%20node%20%5Bshape%3Dbox%20style%3D%22rounded%2Cfilled%22%20fillcolor%3Dgrey90%20fontname%3D%22sans-serif%22%5D%3B%0A%0A%20%20%20%20color%3Dgrey%3B%0A%0A%20%20%20%20cliui%3B%0A%20%20%20%20%22wrap-ansi%22%3B%0A%20%20%20%20yargs%3B%0A%0A%20%20%20%20%7Brank%3Dsame%3B%20cliui%3B%20%22wrap-ansi%22%3B%20yargs%3B%7D%0A%20%20%7D%0A%0A%20%20subgraph%20cluster_1%20%7B%0A%20%20%20%20label%20%3D%20%22Downstream%22%3B%0A%20%20%20%20fontname%3D%22sans-serif%22%3B%0A%20%20%20%20fontcolor%3Dgrey%3B%0A%20%20%20%20shape%3Dbox%3B%0A%20%20%20%20style%3Drounded%3B%0A%0A%20%20%20%20node%20%5Bshape%3Dbox%20style%3D%22rounded%2Cfilled%22%20fillcolor%3Dgrey90%20fontname%3D%22sans-serif%22%5D%3B%0A%20%20%20%20%0A%20%20%20%20color%3Dgrey%3B%0A%0A%20%20%20%20%22emoji-regex%22%3B%0A%20%20%20%20%22is-fullwidth-code-point%22%3B%0A%20%20%20%20%22strip-ansi%22%3B%0A%0A%20%20%20%20%7Brank%3Dsame%3B%20%22emoji-regex%22%3B%20%22is-fullwidth-code-point%22%3B%20%22strip-ansi%22%3B%7D%0A%20%20%7D%0A%0A%20%20%22string-width%22%3B%0A%20%0A%20%20%22cliui%22%20-%3E%20%22string-width%22%3B%0A%20%20%22cliui%22%20-%3E%20%22strip-ansi%22%3B%0A%20%20%22cliui%22%20-%3E%20%22wrap-ansi%22%3B%0A%20%20%22wrap-ansi%22%20-%3E%20%22string-width%22%3B%0A%20%20%22yargs%22%20-%3E%20%22string-width%22%3B%0A%20%20%22string-width%22%20-%3E%20%22emoji-regex%22%3B%0A%20%20%22string-width%22%20-%3E%20%22is-fullwidth-code-point%22%3B%0A%20%20%22string-width%22%20-%3E%20%22strip-ansi%22%3B%0A%20%20%0A%7D)

![image](https://user-images.githubusercontent.com/164050/126070278-10ccf51e-769a-4b01-b371-d86013aba18c.png)

**Comment by fregante** [comment]
I'm not sure I follow, but in that graph why would you show `cliui -> wrap-ansi -> string-width` instead of `cliui -> string-width`? Both are valid.

**Comment by broofa** [comment]
Agreed.  Obviously the primary goal of this view would be to show how the target module is connected to it's upstream / downstream dependencies.  How those dependencies are are inter-connected would only be of secondary-interest.    E.g. The user may not care that `cliui -> strip-ansi` is a thing.

This is just something we'd have to play around with to see how it feels, how users react.  Or make it an option users can configure.

For example, removing edges between dependencies would keep things simple:
![image](https://user-images.githubusercontent.com/164050/126071286-e93cb712-7002-4c58-adca-5cc07fa0c88b.png)

Or, allowing graphviz to layout the upstream / downstream subgraphs:
![image](https://user-images.githubusercontent.com/164050/126071273-7c478baf-c047-4644-8030-505fc74fa95d.png)

**Comment by fregante** [comment]
Closing this RFC issue. You can extract one of your comments into a more specific issue if you'd like to work on it.
